### PR TITLE
[#149410359][#149421881][#149422315][#149422047][#149271291] Orders screen

### PR DESCRIPTION
**Pull request by guillaumedebavelaere**

### CODE REVIEW

File: TradeItIosTicketSDK2/TradeItOrdersTableViewManager.swift [hunk 1]
```diff
@@ -0,0 +1,240 @@
...
+        self.refreshControl = refreshControl
+    }
+    
+    /**
+     * This is to split orders in order to have a specific section for group orders
+    **/
+    private func getSplittedOrdersArray(orders: [TradeItOrderStatusDetails]) -> [[TradeItOrderStatusDetails]]{
+        return orders.reduce([[]], { splittedArrays, order in
+            var splittedArraysTmp = splittedArrays
+            let lastResult: [TradeItOrderStatusDetails] = splittedArrays[(splittedArrays.endIndex - 1)]
```
Comment (jsom) [inline TradeItIosTicketSDK2/TradeItOrdersTableViewManager.swift]: Why not use `.last`?

File: TradeItIosTicketSDK2/TradeItOrdersTableViewManager.swift [hunk 2]
```diff
@@ -0,0 +1,240 @@
...
+        )
+        TradeItThemeConfigurator.configure(view: refreshControl)
+        tableView.addSubview(refreshControl)
+        self.refreshControl = refreshControl
+    }
+    
+    /**
+     * This is to split orders in order to have a specific section for group orders
+    **/
+    private func getSplittedOrdersArray(orders: [TradeItOrderStatusDetails]) -> [[TradeItOrderStatusDetails]]{
```
Comment (jsom) [inline TradeItIosTicketSDK2/TradeItOrdersTableViewManager.swift]: Is there a better name we can think of to use here? I'm not 100% sure what it's doing by something like `groupOrderLegs` or `collateOrders`?

File: TradeItIosTicketSDK2/TradeItOrdersTableViewManager.swift [hunk 3]
```diff
@@ -0,0 +1,240 @@
...
+        return orders.reduce([[]], { splittedArrays, order in
+            var splittedArraysTmp = splittedArrays
+            let lastResult: [TradeItOrderStatusDetails] = splittedArrays[(splittedArrays.endIndex - 1)]
+            
+            if !order.isGroupOrder() && !lastResult.contains(order) { // this is not a group order, we can append the order
+                splittedArraysTmp[(splittedArraysTmp.endIndex - 1)].append(order)
```
Comment (jsom) [inline TradeItIosTicketSDK2/TradeItOrdersTableViewManager.swift]: `.last`

File: TradeItIosTicketSDK2/TradeItOrdersTableViewManager.swift [hunk 4]
```diff
@@ -0,0 +1,240 @@
...
+            self.orderSectionPresenters.append(
+                OrderSectionPresenter(
+                    orders: orders,
+                    title: title,
+                    isGroupOrder: isGroupOrder
+                )
+            )
+        }
+    }
+
```
Comment (jsom) [inline TradeItIosTicketSDK2/TradeItOrdersTableViewManager.swift]: This function is a little bit hard to follow - `forEach` usually indicates that there's is a functional way to do this. Maybe think about moving some of those conditions and `title` decisions to a separate function. Using `map` is probably going to clean this up as well.

File: TradeItIosTicketSDK2/TradeItOrdersViewController.swift [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+        }
+        
+        func ordersPromise() -> Promise<Void> {
+            return Promise<Void> { fulfill, reject in
+                linkedBrokerAccount.getAllOrderStatus(
+                    onSuccess: { orders in
+                        self.ordersTableViewManager?.updateOrders(orders)
+                        fulfill()
+                    },
+                    onFailure: { error in
```
Comment (jsom) [inline TradeItIosTicketSDK2/TradeItOrdersViewController.swift]: `onFailure: reject`

File: TradeItIosTicketSDK2/TradeItOrdersViewController.swift [hunk 6]
```diff
@@ -0,0 +1,94 @@
...
+                    onFailure: reject
+                )
+            }
+        }
+        
+        func ordersPromise() -> Promise<Void> {
+            return Promise<Void> { fulfill, reject in
+                linkedBrokerAccount.getAllOrderStatus(
+                    onSuccess: { orders in
+                        self.ordersTableViewManager?.updateOrders(orders)
```
Comment (jsom) [inline TradeItIosTicketSDK2/TradeItOrdersViewController.swift]: Because this is a side effect rather than the request for data, it might be best to move this below in to its own `then` so set this to `onSuccess: fulfill` and change the promise type to accept the orders. Then below add:
```swift
authenticatePromise().then { _ in
    return ordersPromise()
}.then { orders in
    self.ordersTableViewManager?.updateOrders(orders)
}.catch //... etc
```

**Comment by jsom** [comment]
Looks good @guillaumedebavelaere ! Just a few comments

**Comment by jsom** [comment]
@guillaumedebavelaere are there plans to add a Trade button to the Orders screen?

**Comment by guillaumedebavelaere** [comment]
@jsom No directly,  but there is a trade button when there are no results. Otherwise there is a story to change the portfolio button on the confirmation screen by a button to go to the orders screen.

File: TradeItIosTicketSDK2/TradeItOrdersTableViewManager.swift (hunk 3 above)
Comment (guillaumedebavelaere) [reply TradeItIosTicketSDK2/TradeItOrdersTableViewManager.swift]: Actually last return an immutable variable, so ?I can't use it here, but I will for above

File: TradeItIosTicketSDK2/TradeItOrdersViewController.swift (hunk 6 above)
Comment (jsom) [reply TradeItIosTicketSDK2/TradeItOrdersViewController.swift]: I realise I didn't do this in the recent changes I made to portfolio. I'll try fix it there.

**Comment by jsom** [comment]
Merging this as I'm building a Swift 4 branch
